network.py

```diff
@@ -0,0 +1,32 @@
+import socket
+
+def is_port_available(addr: str, port: int):
+    """
+    Check if a given address:port is available by attempting to create a socket.
+
+    Returns True if the port is available, False otherwise.
+    """
+
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        try:
+            s.bind((addr, port))  # Try to bind the socket to the IP and port
+            return True         # If successful, the port is available
+        except socket.error as e:
+            print(f"Socket error: {e}")
+            return False
+
+def find_available_port(addr: str):
+    """
+    Attempts to find an open port between 1111 and 44444 for the given address.
+
+    Returns the port as a string.
+
+    TODO: implement lease system with etcd for ports - https://etcd.io/docs/v3.4/learning/api/#lease-api
+    """
+
+    # Arbitrarily chose between 1111 and 44444 for ports
+    for i in range(1111, 44444):
+        if is_port_available(addr, i):
+            return i
+
+    return -1
```

Comment:
Probably never happened / rarely but is it possible that two processes determines that a port is available and both take it? If yes, should refactor this logic somehow that when port is returned back, its already binded to the process

Reply:
nit: better error handling

Reply:
> Probably never happened / rarely but is it possible that two processes determines that a port is available and both take it? If yes, should refactor this logic somehow that when port is returned back, its already binded to the process

Shouldn't be an issue most likely. AFAIK if two clients try to bind to a port at the same time, one of them will get rejected

Reply:
> nit: better error handling

Also, this is a temporary thing. etcd has a system where you can "lease" ports which I'm going to implement that soon; this a temporary solution to get everything working

Reply:
> Shouldn't be an issue most likely. AFAIK if two clients try to bind to a port at the same time, one of them will get rejected

Right but that means the returned port from `find_available_port` is unused from one of them right. Unless you have loop to repeatedly check `find_available_port`

Reply:
Yeah. If a given port is taken then it'll try to find another port that's available, so if two threads try to claim a port at the same time, one of them will get rejected and look for another available port.